### PR TITLE
Refactor Result class to be more object oriented

### PR DESCRIPTION
**Pull request by jamieweavis**
Result has been refactored to take a `$matchClass` in the constructor and then the `is*` properties have been converted to methods which use the `$matchClass` to be calculate the result.

Unit tests & classifier have been updated accordingly.

### **_objectorientedification_**

### CODE REVIEW

File: src/Classifier.php [hunk 1]
```diff
@@ -44,17 +44,11 @@ public function classify(Client $parseResult) : Result
...
             }
 
-            $result->isMobileDevice = in_array($rule['class'], ['tablet', 'phone']);
-            $result->isPhone        = $rule['class'] === 'phone';
-            $result->isTablet       = $rule['class'] === 'tablet';
-            $result->isSpider       = $rule['class'] === 'spider';
-            $result->isComputer     = $rule['class'] === 'desktop';
-            $result->matchType      = $rule['class'];
-
+            $matchClass = $rule['class'];
```
Comment (chrisminett) [inline src/Classifier.php]: Feels like the `return` could just be here.

File: src/Classifier.php [hunk 2]
```diff
@@ -44,17 +44,11 @@ public function classify(Client $parseResult) : Result
...
             break;
         }
 
-        return $result;
+        return new Result($matchClass);
```
Comment (chrisminett) [inline src/Classifier.php]: And perhaps this could be an exception? Or is a match class with an empty string valid?

File: src/Result.php [hunk 3]
```diff
@@ -11,32 +11,37 @@
 class Result
 {
     /**
-     * @var string UA-Parser element used for match: 'device', 'os', 'ua'
+     * @var string
      */
-    public $matchType;
+    public $matchClass;
```
Comment (chrisminett) [inline src/Result.php]: `private` ?

File: src/Result.php [hunk 4]
```diff
@@ -11,32 +11,37 @@
...
 
-    /**
-     * @var bool
-     */
-    public $isMobileDevice = false;
+    public function __construct(string $matchClass = '')
+    {
+        $this->matchClass = $matchClass;
```
Comment (chrisminett) [inline src/Result.php]: This could check against an array of allowed types

**Review by chrisminett — COMMENTED** [review]
Not sure whether it's good that it's possible to get a result when there's no match. And possibly if that does change to throw an exception, it would be a separate change to this.

File: src/Result.php (hunk 3 above)
Comment (jamiestraw) [reply src/Result.php]: I thought about this, but it's accessed externally via _ClassifierTest_ on line 54 for the assertion message.

Not sure whether it's worth keeping `public` for this, switching it to `private` and adding a `getMatchClass()` method, or just removing the assertion message altogether.

Comment (chrisminett) [reply src/Result.php]: 👍 for `getMatchClass()`

**Comment by jamiestraw** [comment]
Sorry @chrisminett , didn't see that you had replied - been having problems with my email DNS over the past week... I've actioned the feedback you suggested.

After looking at [this library](https://github.com/matomo-org/device-detector) that James mentioned to me last week after overhearing our Sprint Planning meeting - do you think it's worth considering? I know it would be a shame to throw away the work thus far, but it looks quite good and means that we wouldn't have to maintain our own project 😕

File: src/Result.php [hunk 5]
```diff
@@ -13,13 +13,29 @@ class Result
...
      */
-    public $matchClass;
+    private $matchClass;
+
+    /**
+     * @var array
+     */
+    private $validClasses = ['phone', 'tablet', 'spider', 'computer'];
 
     public function __construct(string $matchClass = '')
```
Comment (chrisminett) [inline src/Result.php]: Seems that there's no need to have a default value / make the param optional

**Review by Pudge601 — APPROVED** [review]

**Review by chrisminett — APPROVED** [review]

**Pull request merged**